packages/client/hmi-client/src/services/workflow.ts

```diff
@@ -634,3 +634,46 @@ export const branchWorkflow = (wf: Workflow, nodeId: string) => {
 	wf.nodes.push(...copyNodes);
 	wf.edges.push(...copyEdges);
 };
+
+function connectOutputOptionToOperator(typeId, key, options: Map<string, Array<{}>>, name) {
+	const displayName: string = name ?? key;
+	if (options.has(typeId)) {
+		const list = options.get(typeId) ?? [];
+		list.push({ type: key, displayName });
+		options.set(typeId, list);
+	} else {
+		options.set(typeId, [{ type: key, displayName }]);
+	}
+}
+
+function getOutputOptions(typeId, key, options: Map<string, Array<string>>) {
+	if (options.has(key)) {
+		const list = options.get(key) ?? [];
+		list.push(typeId);
+		options.set(key, list);
+	} else {
+		options.set(key, [typeId]);
+	}
+}
+
+function getPortOptions(operationMap: Map<string, any>, portKey: string, getOptionsFn: Function) {
+	const portOptions = new Map<string, Array<string>>();
+
+	operationMap.forEach((value, key) => {
+		const port: Array<any> = value?.[portKey] ?? [];
+		port.forEach((option) => getOptionsFn(option.type, key, portOptions, value?.displayName));
+	});
+
+	return portOptions;
+}
```

Comment:
I think you are overcomplicating things with these mutable functions and callbacks, as you have `any` in a handful of places it is harder to follow what is going on.

Since we only have two types of ports, we may consider just spelling it out rather than going the indirect route.